File: tests/st/tensor/mint/test_flatten.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import os
import pytest
import numpy as np
import mindspore as ms
import mindspore.nn as nn
from tests.mark_utils import arg_mark
import random

def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)

class FlattenNet(nn.Cell):
    def construct(self, x, start_dim, end_dim, parm_mode=True):
        if os.environ.get("MS_TENSOR_API_ENABLE_MINT", "0") != '1':
            res = x.flatten(order='C', start_dim=start_dim, end_dim=end_dim)
        elif parm_mode:
            res = x.flatten(start_dim, end_dim)
        else:
            res = x.flatten(start_dim=start_dim, end_dim=end_dim)
        return res

@arg_mark(plat_marks=['platform_ascend'],
          level_mark='level1',
          card_mark='onecard',
          essential_mark='unessential')
@pytest.mark.parametrize("mode", [ms.PYNATIVE_MODE])
def test_flatten_normal(mode):
    """
    Feature: Tensor.flatten.
    Description: Verify the result of flatten.
    Expectation: expect correct result.
    """
    os.environ["MS_TENSOR_API_ENABLE_MINT"] = '1'
    ms.set_context(mode=mode)
    test_shape = (2, 3, 4, 5)
    x = generate_random_input(test_shape, np.float32)
    tx = ms.Tensor(x, dtype=ms.float32)
    net = FlattenNet()
    output = net(tx, 1, 2)
    output2 = net(tx, 1, 2, False)
    expect = x.reshape((2, 12, 5))
    np.testing.assert_allclose(output.asnumpy(), expect, rtol=1e-4)
    np.testing.assert_allclose(output2.asnumpy(), expect, rtol=1e-4)

    output3 = net(ms.Tensor(x), 0, 2)
    output4 = net(ms.Tensor(x), 0, 2, False)
    expect3 = x.reshape((24, 5))
    np.testing.assert_allclose(output3.asnumpy(), expect3, rtol=1e-4)
    np.testing.assert_allclose(output4.asnumpy(), expect3, rtol=1e-4)
    del os.environ["MS_TENSOR_API_ENABLE_MINT"]

@arg_mark(plat_marks=['platform_ascend'],
          level_mark='level1',
          card_mark='onecard',
          essential_mark='unessential')
@pytest.mark.parametrize("mode", [ms.PYNATIVE_MODE])
def test_flatten_random(mode):
    """
    Feature: Tensor.flatten.
    Description: Verify the result of flatten.
    Expectation: expect correct result.
    """
    os.environ["MS_TENSOR_API_ENABLE_MINT"] = '1'
    def flatten_shape(shape, start, end):
        new_shape = []
        for i in range(start):
            new_shape.append(shape[i])
        flattened_dim_size = 1
        for i in range(start, end + 1):
            flattened_dim_size *= shape[i]
        new_shape.append(flattened_dim_size)
        for i in range(end + 1, len(shape)):
            new_shape.append(shape[i])
        return tuple(new_shape)

    ms.set_context(mode=mode)
    test_shape = (2, 3, 4, 5)
    x = generate_random_input(test_shape, np.float32)
    tx = ms.Tensor(x, dtype=ms.float32)
    net = FlattenNet()

    start_dim = [random.randint(0, 3) for _ in range(10)]
    end_dim = [random.randint(0, 3) for _ in range(10)]
    for i in range(10):
        if start_dim[i] > end_dim[i]:
            continue
        output = net(tx, start_dim[i], end_dim[i])
        output2 = net(tx, start_dim[i], end_dim[i], False)
        expect = x.reshape(flatten_shape(test_shape, start_dim[i], end_dim[i]))
        np.testing.assert_allclose(output.asnumpy(), expect, rtol=1e-4)
        np.testing.assert_allclose(output2.asnumpy(), expect, rtol=1e-4)
    del os.environ["MS_TENSOR_API_ENABLE_MINT"]

File: tests/ut/cpp/symbol_engine/ops/test_flatten.cc
/**
 * Copyright 2024 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>

#include "symbol_engine/ops/symbolic_shape_test_utils.h"
#include "common/graph_optimizer_test_framework.h"
#include "abstract/dshape.h"
#include "common/mockcpp.h"

namespace mindspore::symshape::test {
struct FlattenOp {
  ShapeVector shape;
};
struct FlattenExtOp {
  ShapeVector shape;
  int64_t start;
  int64_t end;
};

class TestFlatten : public TestSymbolEngine, public testing::WithParamInterface<FlattenOp> {};
class TestFlattenExt : public TestSymbolEngine, public testing::WithParamInterface<FlattenExtOp> {};

using abstract::TensorShape;
TEST_P(TestFlatten, compare_shape_succ) {
  // building symbolic shape like a dynamic shape node.
  MOCKER_CPP(&TensorShape::IsDynamic, bool (*)(const TensorShape *)).stubs().will(returnValue(true));

  const auto &param = GetParam();
  mindspore::test::ConstructGraph cg;
  auto x = cg.NewTensorInput("x", kFloat32, param.shape);
  auto node = cg.NewCNode("Flatten", {x});
  cg.GetGraph()->set_output(node);
  helper_->InitSymbolEngine(cg.GetGraph());
  auto out_shape = helper_->BuildSymbolicShape(node);
  UT_CHECK_NULL(out_shape);
  ASSERT_TRUE(helper_->SupportInfer());
  ASSERT_TRUE(helper_->CheckSymbolicShapeMatchesDigitalShape(node));
  GlobalMockObject::verify();
}

INSTANTIATE_TEST_CASE_P(TestSymShape, TestFlatten,
                        testing::Values(FlattenOp{{10}},             //
                                        FlattenOp{{10, 20}},         //
                                        FlattenOp{{10, 20, 30}},     //
                                        FlattenOp{{10, 20, 30, 40}}  //
                                        ));

TEST_P(TestFlattenExt, compare_shape_succ) {
  // building symbolic shape like a dynamic shape node.
  MOCKER_CPP(&TensorShape::IsDynamic, bool (*)(const TensorShape *)).stubs().will(returnValue(true));

  const auto &param = GetParam();
  mindspore::test::ConstructGraph cg;
  auto x = cg.NewTensorInput("x", kFloat32, param.shape);
  auto node = cg.NewCNode("FlattenExt", {x, cg.NewValueNode(param.start), cg.NewValueNode(param.end)});
  cg.GetGraph()->set_output(node);
  helper_->InitSymbolEngine(cg.GetGraph());
  auto out_shape = helper_->BuildSymbolicShape(node);
  UT_CHECK_NULL(out_shape);
  ASSERT_TRUE(helper_->SupportInfer());
  ASSERT_TRUE(helper_->CheckSymbolicShapeMatchesDigitalShape(node));
  GlobalMockObject::verify();
}

INSTANTIATE_TEST_CASE_P(TestSymShape, TestFlattenExt,
                        testing::Values(FlattenExtOp{{}, 0, 0},                 //
                                        FlattenExtOp{{10}, 0, -1},              //
                                        FlattenExtOp{{10, 20}, 0, -1},          //
                                        FlattenExtOp{{10, 20, 30}, 0, 1},       //
                                        FlattenExtOp{{10, 20, 30}, 1, 1},       //
                                        FlattenExtOp{{10, 20, 30}, 1, 2},       //
                                        FlattenExtOp{{10, 20, 30, 40}, 0, 2},   //
                                        FlattenExtOp{{10, 20, 30, 40}, 0, -1},  //
                                        FlattenExtOp{{10, 20, 30, 40}, 1, 2},   //
                                        FlattenExtOp{{10, 20, 30, 40}, -3, -1}  //
                                        ));
}  // namespace mindspore::symshape::test
